File: frontend/src/api/mccInstance.ts
import client from './client'

export interface MccInstanceCreatePayload {
  slug: string
  display_name?: string
  bot_id?: string | null
  account_profile_id?: string | null
  binary_mode?: 'symlink' | 'copy' | 'external'
  mc_username?: string
  mc_server_host?: string
  mc_server_port?: number
  mc_version?: string
}

export interface MccInstance {
  instance_id: string
  slug: string
  display_name: string
  bot_id: string | null
  account_profile_id: string | null
  instance_dir: string
  binary_mode: string
  mcc_binary_path: string
  status: string
  desired_state: string
  pid: number | null
  exit_code: number | null
  mcp_host: string
  mcp_port: number
  mcp_auth_token_env: string
  mc_username: string
  mc_server_host: string
  mc_server_port: number
  mc_version: string
  organization_id: string | null
  created_by: string | null
  last_started_at: string | null
  last_stopped_at: string | null
  last_heartbeat_at: string | null
  created_at: string
  updated_at: string
}

export interface MccTerminalLine {
  seq: number
  stream: string
  content: string
  created_at: string
}

export interface MccFileEntry {
  name: string
  path: string
  type: 'directory' | 'file'
  size: number
  updated_at: number
  editable: boolean
  downloadable: boolean
  language: string
}

export interface MccFileTreeNode {
  name: string
  path: string
  type: 'directory' | 'file'
  children: MccFileTreeNode[]
}

export interface MccFileBreadcrumb {
  name: string
  path: string
}

export interface MccFileContent {
  path: string
  content: string
  encoding: string
  size: number
  language: string
  masked: boolean
  updated_at: number
}

export interface MccFileSaveResponse {
  path: string
  size?: number | null
  snapshot_id?: string | null
  diff?: string | null
  masked_secrets_preserved?: boolean
}

export interface MccFileDownloadResponse {
  path: string
  name: string
  size: number
  content_base64: string
  language: string
  download_name: string
}

export type MccAuthType = 'offline' | 'microsoft' | 'mojang' | 'yggdrasil' | 'custom'

export interface MccAccountConfig {
  auth_type: MccAuthType
  username: string
  password_set: boolean
  password?: string | null
  auth_server_url: string
  auth_api_path: string
  authlib_injector_path: string
  mc_server_host: string
  mc_server_port: number
  mc_version: string
  mcp_port: number
  mcp_auth_token_env: string
}

export interface MccAccountProfile {
  profile_id: string
  name: string
  auth_type: MccAuthType
  username: string
  password_set: boolean
  password?: string | null
  auth_server_url: string | null
  auth_api_path: string | null
  authlib_injector_path: string | null
  mc_server_host: string
  mc_server_port: number
  mc_version: string
  last_login_name: string | null
  organization_id: string | null
  created_at: string
  updated_at: string
}

export interface MccAccountProfilePayload {
  name: string
  auth_type: MccAuthType
  username: string
  password?: string | null
  auth_server_url?: string
  auth_api_path?: string
  authlib_injector_path?: string
  mc_server_host?: string
  mc_server_port?: number
  mc_version?: string
}

export const mccInstanceApi = {
  listProfiles() {
    return client.get<{ items: MccAccountProfile[]; total: number }>('/mcc/instances/account-profiles')
  },
  createProfile(data: MccAccountProfilePayload) {
    return client.post<MccAccountProfile>('/mcc/instances/account-profiles', data)
  },
  updateProfile(profileId: string, data: Partial<MccAccountProfilePayload> & { clear_password?: boolean }) {
    return client.patch<MccAccountProfile>(`/mcc/instances/account-profiles/${profileId}`, data)
  },
  deleteProfile(profileId: string) {
    return client.delete(`/mcc/instances/account-profiles/${profileId}`)
  },
  applyProfile(instanceId: string, profileId: string) {
    return client.post<{ config: MccAccountConfig; snapshot_id: string; diff: string; restart_required: boolean }>(`/mcc/instances/${instanceId}/account-config/apply-profile`, { profile_id: profileId })
  },
  list(status?: string) {
    return client.get<{ items: MccInstance[]; total: number }>('/mcc/instances', { params: { status } })
  },
  create(data: MccInstanceCreatePayload) {
    return client.post<MccInstance>('/mcc/instances', data)
  },
  get(instanceId: string) {
    return client.get<MccInstance>(`/mcc/instances/${instanceId}`)
  },
  update(instanceId: string, data: Partial<MccInstanceCreatePayload>) {
    return client.patch<MccInstance>(`/mcc/instances/${instanceId}`, data)
  },
  delete(instanceId: string) {
    return client.delete(`/mcc/instances/${instanceId}`)
  },
  start(instanceId: string, env: Record<string, string> = {}) {
    return client.post(`/mcc/instances/${instanceId}/start`, { env })
  },
  stop(instanceId: string, force = false, timeout_seconds = 10) {
    return client.post(`/mcc/instances/${instanceId}/stop`, { force, timeout_seconds })
  },
  restart(instanceId: string) {
    return client.post(`/mcc/instances/${instanceId}/restart`)
  },
  history(instanceId: string, tail = 200) {
    return client.get<{ items: MccTerminalLine[]; last_seq: number }>(`/mcc/instances/${instanceId}/terminal/history`, { params: { tail } })
  },
  input(instanceId: string, input: string, append_newline = true) {
    return client.post(`/mcc/instances/${instanceId}/terminal/input`, { input, append_newline })
  },
  listFiles(instanceId: string, path = '') {
    return client.get<{ path: string; breadcrumbs: MccFileBreadcrumb[]; items: MccFileEntry[] }>(`/mcc/instances/${instanceId}/files`, { params: { path } })
  },
  listFileTree(instanceId: string, path = '') {
    return client.get<{ items: MccFileTreeNode[] }>(`/mcc/instances/${instanceId}/files/tree`, { params: { path } })
  },
  downloadFile(instanceId: string, path: string) {
    return client.get<MccFileDownloadResponse>(`/mcc/instances/${instanceId}/files/download`, { params: { path } })
  },
  readFile(instanceId: string, path: string) {
    return client.get<MccFileContent>(`/mcc/instances/${instanceId}/files/content`, { params: { path } })
  },
  saveFile(instanceId: string, path: string, content: string, encoding = 'utf-8') {
    return client.put<MccFileSaveResponse>(`/mcc/instances/${instanceId}/files/content`, { path, content, encoding })
  },
  createFile(instanceId: string, path: string, content = '', overwrite = false) {
    return client.post<MccFileSaveResponse>(`/mcc/instances/${instanceId}/files`, { path, content, overwrite })
  },
  createDirectory(instanceId: string, path: string, overwrite = false) {
    return client.post<{ path: string; type: string }>(`/mcc/instances/${instanceId}/directories`, { path, overwrite })
  },
  uploadFile(instanceId: string, path: string, contentBase64: string, overwrite = false) {
    return client.post<MccFileSaveResponse>(`/mcc/instances/${instanceId}/files/upload`, { path, content: contentBase64, overwrite })
  },
  deleteFile(instanceId: string, path: string) {
    return client.delete(`/mcc/instances/${instanceId}/files`, { params: { path } })
  },
  renameFile(instanceId: string, source_path: string, target_path: string, overwrite = false) {
    return client.post(`/mcc/instances/${instanceId}/files/rename`, { source_path, target_path, overwrite })
  },
  getAccountConfig(instanceId: string) {
    return client.get<MccAccountConfig>(`/mcc/instances/${instanceId}/account-config`)
  },
  saveAccountConfig(instanceId: string, config: MccAccountConfig) {
    return client.put<{ config: MccAccountConfig; snapshot_id: string; diff: string; restart_required: boolean }>(`/mcc/instances/${instanceId}/account-config`, config)
  },
}

File: vmtools-next/src/vmtools_next/api/schemas/mcc_instance.py
"""Schemas for MCC remote instance management."""
from __future__ import annotations

from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, Field, field_validator


class MccInstanceCreate(BaseModel):
    slug: str = Field(min_length=1, max_length=64)
    display_name: str = ""
    bot_id: Optional[str] = None
    account_profile_id: Optional[str] = None
    binary_mode: str = "symlink"
    mc_username: str = ""
    mc_server_host: str = ""
    mc_server_port: int = 25565
    mc_version: str = "1.21.1"

    @field_validator("slug")
    @classmethod
    def validate_slug(cls, value: str) -> str:
        import re

        normalized = value.strip().lower()
        if not re.fullmatch(r"[a-z0-9][a-z0-9_-]{0,63}", normalized):
            raise ValueError("slug must match [a-z0-9][a-z0-9_-]{0,63}")
        return normalized

    @field_validator("binary_mode")
    @classmethod
    def validate_binary_mode(cls, value: str) -> str:
        if value not in {"symlink", "copy", "external"}:
            raise ValueError("binary_mode must be symlink, copy, or external")
        return value


class MccInstanceUpdate(BaseModel):
    display_name: Optional[str] = None
    bot_id: Optional[str] = None
    account_profile_id: Optional[str] = None
    mc_username: Optional[str] = None
    mc_server_host: Optional[str] = None
    mc_server_port: Optional[int] = None
    mc_version: Optional[str] = None


class MccInstanceStartRequest(BaseModel):
    env: dict[str, str] = Field(default_factory=dict)


class MccInstanceStopRequest(BaseModel):
    force: bool = False
    timeout_seconds: float = 10.0


class MccTerminalInputRequest(BaseModel):
    input: str = Field(min_length=1, max_length=4096)
    append_newline: bool = True


class MccFileEntryResponse(BaseModel):
    name: str
    path: str
    type: str
    size: int
    updated_at: float
    editable: bool
    downloadable: bool = True
    language: str = ""


class MccFileListResponse(BaseModel):
    path: str
    breadcrumbs: list[dict[str, str]] = Field(default_factory=list)
    items: list[MccFileEntryResponse]


class MccFileTreeNode(BaseModel):
    name: str
    path: str
    type: str
    children: list["MccFileTreeNode"] = Field(default_factory=list)


class MccFileTreeResponse(BaseModel):
    items: list[MccFileTreeNode]


class MccFileContentResponse(BaseModel):
    path: str
    content: str
    encoding: str
    size: int
    language: str
    masked: bool
    updated_at: float


class MccFileWriteRequest(BaseModel):
    path: str = Field(min_length=1, max_length=512)
    content: str = Field(max_length=1_048_576)
    encoding: str = "utf-8"


class MccFileCreateRequest(BaseModel):
    path: str = Field(min_length=1, max_length=512)
    content: str = Field(default="", max_length=1_398_104)
    encoding: str = "utf-8"
    overwrite: bool = False


class MccDirectoryCreateRequest(BaseModel):
    path: str = Field(min_length=1, max_length=512)
    overwrite: bool = False


class MccFileRenameRequest(BaseModel):
    source_path: str = Field(min_length=1, max_length=512)
    target_path: str = Field(min_length=1, max_length=512)
    overwrite: bool = False


class MccFileSaveResponse(BaseModel):
    path: str
    size: int | None = None
    snapshot_id: str | None = None
    diff: str | None = None
    masked_secrets_preserved: bool = False


class MccAccountConfigResponse(BaseModel):
    auth_type: str
    username: str
    password_set: bool
    auth_server_url: str = ""
    auth_api_path: str = ""
    authlib_injector_path: str = ""
    mc_server_host: str
    mc_server_port: int
    mc_version: str
    mcp_port: int
    mcp_auth_token_env: str


class MccAccountConfigUpdate(BaseModel):
    auth_type: str = Field(default="offline")
    username: str = Field(min_length=1, max_length=128)
    password: Optional[str] = Field(default=None, max_length=512)
    auth_server_url: str = ""
    auth_api_path: str = ""
    authlib_injector_path: str = ""
    mc_server_host: str = ""
    mc_server_port: int = Field(default=25565, ge=1, le=65535)
    mc_version: str = "1.21.1"

    @field_validator("auth_type")
    @classmethod
    def validate_auth_type(cls, value: str) -> str:
        if value not in {"offline", "microsoft", "mojang", "yggdrasil", "custom"}:
            raise ValueError("auth_type must be offline, microsoft, mojang, yggdrasil, or custom")
        return value


class MccAccountConfigSaveResponse(BaseModel):
    config: MccAccountConfigResponse
    snapshot_id: str
    diff: str
    restart_required: bool = True


class MccAccountProfileCreate(BaseModel):
    name: str = Field(min_length=1, max_length=128)
    auth_type: str = "offline"
    username: str = Field(min_length=1, max_length=128)
    password: Optional[str] = Field(default=None, max_length=512)
    auth_server_url: str = ""
    auth_api_path: str = ""
    authlib_injector_path: str = ""
    mc_server_host: str = ""
    mc_server_port: int = Field(default=25565, ge=1, le=65535)
    mc_version: str = "1.21.1"

    @field_validator("auth_type")
    @classmethod
    def validate_auth_type(cls, value: str) -> str:
        if value not in {"offline", "microsoft", "mojang", "yggdrasil", "custom"}:
            raise ValueError("auth_type must be offline, microsoft, mojang, yggdrasil, or custom")
        return value


class MccAccountProfileUpdate(BaseModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=128)
    auth_type: Optional[str] = None
    username: Optional[str] = Field(default=None, min_length=1, max_length=128)
    password: Optional[str] = Field(default=None, max_length=512)
    clear_password: bool = False
    auth_server_url: Optional[str] = None
    auth_api_path: Optional[str] = None
    authlib_injector_path: Optional[str] = None
    mc_server_host: Optional[str] = None
    mc_server_port: Optional[int] = Field(default=None, ge=1, le=65535)
    mc_version: Optional[str] = None

    @field_validator("auth_type")
    @classmethod
    def validate_auth_type(cls, value: Optional[str]) -> Optional[str]:
        if value is not None and value not in {"offline", "microsoft", "mojang", "yggdrasil", "custom"}:
            raise ValueError("auth_type must be offline, microsoft, mojang, yggdrasil, or custom")
        return value


class MccAccountProfileResponse(BaseModel):
    profile_id: str
    name: str
    auth_type: str
    username: str
    password_set: bool
    auth_server_url: str | None = None
    auth_api_path: str | None = None
    authlib_injector_path: str | None = None
    mc_server_host: str
    mc_server_port: int
    mc_version: str
    last_login_name: str | None = None
    organization_id: str | None = None
    created_at: datetime
    updated_at: datetime


class MccAccountProfileListResponse(BaseModel):
    items: list[MccAccountProfileResponse]
    total: int


class MccApplyAccountProfileRequest(BaseModel):
    profile_id: str


class MccInstanceResponse(BaseModel):
    instance_id: str
    slug: str
    display_name: str
    bot_id: Optional[str] = None
    account_profile_id: Optional[str] = None
    instance_dir: str
    binary_mode: str
    mcc_binary_path: str
    status: str
    desired_state: str
    pid: Optional[int] = None
    exit_code: Optional[int] = None
    mcp_host: str
    mcp_port: int
    mcp_auth_token_env: str
    mc_username: str
    mc_server_host: str
    mc_server_port: int
    mc_version: str
    organization_id: Optional[str] = None
    created_by: Optional[str] = None
    last_started_at: Optional[datetime] = None
    last_stopped_at: Optional[datetime] = None
    last_heartbeat_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class MccInstanceListResponse(BaseModel):
    items: list[MccInstanceResponse]
    total: int


class MccTerminalLogResponse(BaseModel):
    seq: int
    stream: str
    content: str
    created_at: datetime


class MccTerminalHistoryResponse(BaseModel):
    items: list[MccTerminalLogResponse]
    last_seq: int


class MccProcessEventResponse(BaseModel):
    event_id: int
    instance_id: str
    event_type: str
    pid: Optional[int] = None
    exit_code: Optional[int] = None
    message: str
    created_at: datetime

    model_config = {"from_attributes": True}


class MccStartStopResponse(BaseModel):
    instance_id: str
    status: str
    pid: Optional[int] = None
    mcp_port: int
    message: str = ""
    extra: dict[str, Any] = Field(default_factory=dict)
